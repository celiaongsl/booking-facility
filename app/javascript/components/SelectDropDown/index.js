import React from "react";

import { makeStyles } from "@material-ui/core/styles";

import Select from "@material-ui/core/Select";
import MenuItem from "@material-ui/core/MenuItem";
import InputLabel from "@material-ui/core/InputLabel";
import FormControl from "@material-ui/core/FormControl";

const useStyles = makeStyles({
  formControl: {
    margin: 10,
    minWidth: 120,
  },
});

const SelectDropDown = (props) => {
  const classes = useStyles();
  const { label, onChange, value, list } = props;

  return (
    <FormControl className={classes.formControl}>
      <InputLabel id="demo-simple-select-label">{label}</InputLabel>
      <Select
        labelId="demo-simple-select-label"
        id="demo-simple-select"
        value={value}
        onChange={(e) => onChange(e.target.value)}
        inputProps={{ "data-testid": "select" }}
      >
        {list.map((item) => {
          return (
            <MenuItem
              key={item.value}
              value={item.value}
            >
              {item.label}
            </MenuItem>
          );
        })}
      </Select>
    </FormControl>
  );
};

export default SelectDropDown;
