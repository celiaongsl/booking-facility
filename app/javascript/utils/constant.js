import React from 'react';

import TvRoundedIcon from "@material-ui/icons/TvRounded";
import PowerRoundedIcon from '@material-ui/icons/PowerRounded';
import NoteOutlinedIcon from '@material-ui/icons/NoteOutlined';
import CastRoundedIcon from '@material-ui/icons/CastRounded';

export const assetToIconList = {
  tv: {
    icon: <TvRoundedIcon />,
    label: "TV",
  },
  power_outlets: {
    label: "Power Outlets",
    icon: <PowerRoundedIcon />
  },
  whiteboard: {
    label: "Whiteboard",
    icon: <NoteOutlinedIcon />
  },
  projector: {
    label: "Projector",
    icon: <CastRoundedIcon />
  },
};

export const apiURL = '/api/v1';