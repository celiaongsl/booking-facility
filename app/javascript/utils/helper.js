import { assetToIconList } from "./constant";

export const findAssetKeyReturnIconList = (assets) => {
  let assetsIconList = [];
  Object.keys(assets).map((key) => {
    if (assets[key]) {
      return assetsIconList.push(assetToIconList[key].icon);
    }
    return null;
  });
  return assetsIconList;
};

export const roundTimeQuarterHour = (time) => {
  var timeToReturn = new Date(time);

  timeToReturn.setMilliseconds(
    Math.round(timeToReturn.getMilliseconds() / 1000) * 1000
  );
  timeToReturn.setSeconds(Math.round(timeToReturn.getSeconds() / 60) * 60);
  timeToReturn.setMinutes(Math.round(timeToReturn.getMinutes() / 15) * 15);
  return timeToReturn;
};

// Take "Thu Feb 18 2021 19:15:00 GMT+0800 (Singapore Standard Time)"
// And shorten to "19:15:00" only
export function prettyDate2(time) {
  if (typeof time === "string") time = new Date(time);
  var localeSpecificTime = time.toLocaleTimeString();
  return localeSpecificTime.replace(/:\d+ /, " ");
}

export const convertObjectIntoQueryString = ({ object, paramKey }) => {
  // if capacity[key] === true, we add capacity[]=key in the query
  // so if we have multiple capacity keys that are true, we need to chain it as such:
  // ?capacity[]=1&capacity[]=30 << return this back

  let queryString = "";
  Object.keys(object).map((key) => {
    if (object[key]) {
      queryString = queryString + `&${paramKey}=${key}`;
    }
  });

  return queryString;
};
