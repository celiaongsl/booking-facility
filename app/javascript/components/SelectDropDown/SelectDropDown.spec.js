import React from "react";
import SelectDropDown from "./";
import { render, fireEvent } from "@testing-library/react";
import { time } from "../../utils/time_constant";

const props = {
  label: "Start Time",
  onChange: jest.fn(),
  value: "00:15:00",
  list: time,
};

describe("<SelectDropDown />", () => {
  it("renders the component", () => {
    const { asFragment } = render(<SelectDropDown {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should be able to mount the select component", async () => {
    const value = "00:15:00";
    let { container } = render(<SelectDropDown {...props} />);

    expect(container.querySelector("input")).toHaveProperty("value", value);
  });

  [" ", "ArrowUp", "ArrowDown", "Enter"].forEach((key) => {
    it(`should open menu when pressed ${key} key on select`, () => {
      const { getByRole, getAllByRole } = render(<SelectDropDown {...props} />);
      const trigger = getByRole("button");
      trigger.focus();

      fireEvent.keyDown(trigger, { key });
      expect(getByRole("listbox", { hidden: false })).not.toBeNull();

      fireEvent.keyUp(getAllByRole("option")[0], { key });
      expect(getByRole("listbox", { hidden: false })).not.toBeNull();
    });
  });

  it("should not be called if selected element has the current value (value did not change)", () => {
    const { getAllByRole, getByRole } = render(<SelectDropDown {...props} />);
    fireEvent.mouseDown(getByRole("button"));
    getAllByRole("option")[1].click();
    expect(props.onChange).toBeCalledTimes(0);
  });

  it("should be called if selected element does not have the current value (value changed)", () => {
    const { getAllByRole, getByRole } = render(<SelectDropDown {...props} />);
    fireEvent.mouseDown(getByRole("button"));
    getAllByRole("option")[3].click();
    expect(props.onChange).toBeCalledTimes(1);
  });
});
