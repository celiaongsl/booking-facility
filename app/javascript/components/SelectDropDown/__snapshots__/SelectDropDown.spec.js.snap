// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectDropDown /> renders the component 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root makeStyles-formControl-1"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
      data-shrink="true"
      id="demo-simple-select-label"
    >
      Start Time
    </label>
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
    >
      <div
        aria-haspopup="listbox"
        aria-labelledby="demo-simple-select-label demo-simple-select"
        class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
        id="demo-simple-select"
        role="button"
        tabindex="0"
      >
        12:15 AM
      </div>
      <input
        aria-hidden="true"
        class="MuiSelect-nativeInput"
        data-testid="select"
        tabindex="-1"
        value="00:15:00"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSelect-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
